---
title: "Animated Manhattan Plot"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Manhattan and QQ plots are used to interpret the results of genome-wide association studies (GWASs). Here we build animated gifs based on multiple Manhattan or QQ plots.

## Coloring based on effect on pathways

In this section, we color the Manhattan based on possible effect on pathways. As an example, We are going to use the results of association with BMI from the GIANT consortium. Note that here we retain only SNPs with an rs ID.

```{r input}

giantData <- read.table(file = "resources/bmi_p-values.gz", header = T, stringsAsFactors = F, sep = "\t")
giantData <- giantData[giantData$SNPNAME != '-', ]

```

### Pathway matching

In order to speculate on the effect of variants on pathways, the effect of genetic variants on proteins is first inferred using Ensembl Variant Effect Predictor, [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html), and then the participation of proteins in pathways is inferred using [Reactome](https://reactome.org). This analysis can readily be conducted using [PathwayMatcher](https://github.com/PathwayAnalysisPlatform/PathwayMatcher).

:pencil2: Extract the ids of the variants to match in a single file.

```{r input}

writeLines(giantData$SNPNAME, "snps.txt")

```

:pencil2: If it is not done already, install PathwayMatcher. The easiest is to do it using Bioconda.

```

conda install -c bioconda pathwaymatcher

```

:pencil2: Run PathwayMatcher on the snp list produced previously. Note that the command below is for use _via_ bioconda.

```

pathwaymatcher -i snps.txt -t rsid -o outputFolder -tlp

```

PathwayMatcher should produce two files named _analysis.tsv_ and _search.tsv_. The search file contains the snp to pathway mapping. Note that these files can be found in the _resources_ folder.

