---
title: "Basic Metods for Genome Wide Association"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this tutorial, we will introduce basic solutions to manipulate and visualize the results of genome wide sequencing data. For a complete protocol please refer to [(1)](#references).

## Installation

We will need the following libraries: _ggplot2_ and _qqman_. Make sure they are installed in your R session.

```{r libraries}
install.packages("ggplot2", repos="http://cran.uib.no/")
install.packages("qqman", repos="http://cran.uib.no/")
```


```{r libraries}
library(ggplot2)
library(qqman)
```


## Admixture

A principal component analysis can be used to visualize the main source of variance in the dataset. Please download the file [pca.txt]() to your working folder.

```{r pca_input}
pcaInput <- read.table(file = "pca.txt", header = T, stringsAsFactors = F)
```

This file contains the results of a PCA conducted on the thousands genomes with associated ethnicity. The variance might be explained by admixture in the dataset. 

```{r pca_plot}
pcaPlot <- ggplot()
pcaPlot <- pcaPlot + geom_point(aes(x=pcaInput$pca1, y=pcaInput$pca2, col = pcaInput$population), alpha = 0.5)
pcaPlot <- pcaPlot + labs(x = "PC1", y = "PC2", col = "Population")
plot(pcaPlot)
```

It is possible to set thresholds to prune the population down to a core population by removing outliers on the different dimensions of the PCA.


## QQ and Manhattan Plots

Please download the file [association.gz]() to your working folder. Note that the file is compressed, but you will be able to read it like any file using the _gzip_ function.

```{r pca_input}
associationInput <- read.table(gzfile("association.gz"), header = T, stringsAsFactors = F)
```

Note that for the sake of time, the dataset has been limited to 5,000,000 markers randomly chosen.

We are going to inspect the calibration of the p-values by drawing a qq-plot. We also estimate lambda that will be image of the inflation at the mediane.

```{r qq_plot}
pValues <- associationInput$P
chisq <- qchisq(1-pValues,1)
lambda <- median(chisq)/qchisq(0.5,1)

expectedPValues <- -log10(chisq)
observedPValues <- -log10(pValues)

qqPlot <- ggplot()
qqPlot <- qqPlot + geom_point(aes(x=expectedPValues, y=observedPValues), size = 1, col = "blue")
qqPlot <- qqPlot + geom_line(aes(x=observedPValues, y=observedPValues), size = 1, alpha = 0.5, linetype = "dotted")
qqPlot <- qqPlot + xlab("Expected p-value [-log]") + ylab("Observed p-value [-log]")
plot(qqPlot)

paste("lambda: ", lambda, sep = "")
```

The inspection of associated SNPs on the whole genome is done using a so-called Manhattan plot. The method is readily implemented in the _qqman_ package. Note that its execution can take time depending on the number of markers and the performance of the computer.

```{r manhattan_plot}
manhattan(associationInput)
```


## References

(1) [Clarke, GM. _et al._, Basic statistical analysis in genetic case-control studies, Nature Protocols, 2011](https://www.ncbi.nlm.nih.gov/pubmed/21293453)

